### ISSUE TITLE
Shallow copying of array in Tables

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1143]
**Reported by sjoelund&#46;se on 29 May 2013 07:02 UTC**
From #1138:

> ModelicaStandardTables.c

```c
279:        tableID->cols = cols;
301:        tableID->table = table;
```

The above assignments simply assign work arrays to the external object. This is a shallow copy that may not be valid once the constructor returns. Thus, using these struct members in subsequent functions gives undefined result.

They need be changed to use malloc+copy+free in order to be safe in any tool.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1143

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 07:07 UTC**
Actually `tableID->cols` is not size_t\*, so it should have worked. Anyway, I will soon provide you a patch for OM x64 test.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 29 May 2013 07:14 UTC**
Yes, that is not the real issue. The issue is that OpenModelica internally uses long, and Modelica external uses int. So we have, say (and I paraphrase):

```
long longcols[16] = {...};
omc_ext_fn(longcols, ...);
```

Where we do:
```
int intcols[16]; // need to convert to Modelica external C array
for (...) intcols[i] = longcols[i];
ext_fn(intcols,...); // Calling your code
// intcols are free'd. anything using them is pointing to garbage
```

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 07:29 UTC**
Please see attachment:T1143.patch. OM x64 should hopefully no longer crash.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 29 May 2013 07:36 UTC**
The patch has tabs (which I despise as the formatting is now wrong). More importantly, the table data is not copied. But I will test it nevertheless.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 07:41 UTC**
New attachment:T1143.patch comes with leading blanks. table array is still not copied.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 29 May 2013 07:47 UTC**
Yes, with the patch there is no longer a crash. I now need to verify if the result is correct without optimisation.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 29 May 2013 08:01 UTC**
Replying to [comment:6 sjoelund.se]:
> Yes, with the patch there is no longer a crash. I now need to verify if the result is correct without optimisation.

The result is not correct. You need to also copy the table data.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 08:07 UTC**
What kind of optimisation did you disable?

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 29 May 2013 08:35 UTC**
None. We do not optimise parameter matrices (only constants, and probably 1-D parameter vectors).

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 May 2013 08:53 UTC**
I will wait for lib officers decision on how to solve this.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 29 May 2013 13:46 UTC**
It seems quite odd that due to the implementation in one tool, all other tools need to copy a lot of data (I now understand also the issue with "col": OpenModelica has a different representation of Integer arrays internally (64 bit) and in the external function interface (32 bit) and therefore OpenModelica has to copy Integer arrays for external function calls and therefore the pointer has only a short lifetime).

Therefore, I propose the following:

  Add a "define" flag to the table implementation and when this flag is set, then all pointer data is copied.
  If this flag is not set (= the default), then pointer data is not copied.

By this, all tools can get an appropriate implementation.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 29 May 2013 13:58 UTC**
otter: There is a real bug here. The Modelica specification does not allow you to use the inputs in the way the new tables does (nothing is said about the programming language a Modelica tool should use: the code assumes a certain C implementation). The default for sure should be to copy the pointer data. Otherwise you will get angry users in a while when more complicated use cases pop up. I am certain of this.

The input is declared as a variable, but it is used under the assumption that all inputs are constant. Also, there is very little in the specification regarding external object constructors anyway. Probably the only sane way to define them is to force constants to be used (or function inputs if it is a local object). Here the input is parameter.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 29 May 2013 14:18 UTC**
Replying to [comment:12 sjoelund.se]:
> otter: There is a real bug here. The Modelica specification does not allow you to use the inputs in the way the new tables does (nothing is said about the programming language a Modelica tool should use: the code assumes a certain C implementation). The default for sure should be to copy the pointer data. Otherwise you will get angry users in a while when more complicated use cases pop up. I am certain of this.

O.k. Fine: Default is "copy", but with a "define" it is possible to not copy.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 31 May 2013 13:32 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 31 May 2013 18:57 UTC**
Resolved by c4eba6d2814ce4ea40d96cb8c55e6c86e0b2fe4d on source level (after phone call with Martin Otter).

There is a new define `NO_TABLE_COPY` to keep the table pointer and do not copy the table array.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 2 Jun 2013 19:14 UTC**
